scarb/src/compiler/db.rs

```diff
@@ -30,7 +34,49 @@ pub(crate) fn build_scarb_root_database(
         }
     }
 
-    b.build()
+    let mut db = b.build()?;
+    inject_virtual_wrapper_lib(&mut db, unit)?;
+    Ok(db)
+}
+
+/// Generates a wrapper lib file for appropriate compilation units.
+///
+/// This approach allows compiling crates that do not define `lib.cairo` file.
+/// For example, single file crates can be created this way.
+/// The actual single file module is defined as `mod` item in created lib file.
+fn inject_virtual_wrapper_lib(db: &mut RootDatabase, unit: &CompilationUnit) -> Result<()> {
+    let components: Vec<&CompilationUnitComponent> = unit
+        .components
+        .iter()
+        .filter(|component| !component.package.id.is_core())
+        // Skip components defining the default source path, as they already define lib.cairo files.
+        .filter(|component| {
+            component
+                .target
+                .source_path
+                .file_name()
+                .map(|file_name| file_name != DEFAULT_MODULE_MAIN_FILE)
+                .unwrap_or(false)
+        })
+        .collect();
+
+    for component in components {
+        let crate_name = component.cairo_package_name();
+        let crate_id = db.intern_crate(CrateLongId::Real(crate_name));
+        let file_stem = component.target.source_path.file_stem().ok_or_else(|| {
```

Comment:
hmm, how this will react when we create a file `tests/ęśąćż.cairo` or `tests/file space.cairo`? both are not valid identifiers, are we validating these?

not a topic to fix in this PR, I'd rather create a bug report labelled as good first issue 😄